### DISCUSSION TITLE
Compound Functional Unit

### DISCUSSION THREAD

**Discussion by atg-abhishek** [Green-Software-Foundation/sci#238] [SCI feedback]
Based on the case study proposed by @srini1978 and discussed by @jawache in [this GDoc](https://docs.google.com/document/d/1UI1UpLleF9vII-_KxAUP1t7kaPZJ4LJD0hNaoMfbGr8/edit?usp=sharing), it will be interesting to analyze the implications of having a **compound functional unit**, something that is not just a single unit (like number of users), but the number of users concurrently who are serviced over a particular time period. E.g. 500 Users over a 1 hour period with 50 concurrent. 

The first implication that comes to mind is how difficult it would make comparability when others try to do these calculations. Any other implications that come to mind?

**Comment by srini1978** [comment]
> Based on the case study proposed by @srini1978 and discussed by @jawache in [this GDoc](https://docs.google.com/document/d/1UI1UpLleF9vII-_KxAUP1t7kaPZJ4LJD0hNaoMfbGr8/edit?usp=sharing), it will be interesting to analyze the implications of having a **compound functional unit**, something that is not just a single unit (like number of users), but the number of users concurrently who are serviced over a particular time period. E.g. 500 Users over a 1 hour period with 50 concurrent.
> 
> The first implication that comes to mind is how difficult it would make comparability when others try to do these calculations. Any other implications that come to mind?

This is a great question @atg-abhishek  . 

For me, measuring SCI score is all about how the SCI score and it's component's individual values are going down post the tunings we undertake for either hardware efficiency, energy efficiency and/or carbon awareness.

If we take that perspective, then we are interested in how the software is operating and how we are measuring these values.  In that context, my stand is that we will never be able to compare SCI values across applications and it will always be contextual.  For example a company builds a time reporting web application using Python for it's 50,000 users. The earlier app was running in maybe Visual basic for the same number of users.  We can compare the SCI score for the same app across 2 different technology stacks and call out that maybe the Python app is more "green". 

But we will not be able to compare the SCI score of this time reporting application with "candy crush" app.  

Why ? Because the context is very different . Then comes the question if we need to bring this context quotient into our SCI formula. My answer would be that it is probably difficult to do that and that we will have to leave it the SCI formula user to be able to decipher the context and figure out the "context" in which to compare SCI scores. 

@jawache  @Henry-WattTime  Tagging you as well for thoughts.

**Reply by Henry-WattTime** [reply]
Philosophically, I currently agree with @srini1978 that it will be difficult to make SCI scores comparable/stackable. I think this is something that the working group will have to extensively grapple with any may be a v2 project due to the complexity. 

Of course, I've always been a 'this is a tool for developers to make their software greener' advocate, but that is just one of the goals of the standard. Comparability is also a goal, but until now we've shunted that down the road.

To actually answer the question, yes, I think compound functional units will be necessary. @Tammy and I encountered this same issue when calculating a score for her pi-based sensor array on her farm. I think we finally landed on a sensor array year as the functional unit:
https://github.com/Green-Software-Foundation/software_carbon_intensity/issues/241#issuecomment-1068378777

**Reply by atg-abhishek** [reply]
@srini1978 thanks for your reply, I guess to clarify when talking about comparability, I was thinking about it in terms of the same class of applications, so productivity apps -> time reporting with similar apps rather than against games. The reason being that it would help an end-user make an informed choice to pick Solution X over Solution Y, when both provide similar functionality, towards the goal of being greener in their consumption.

**Reply by ElmarBorgmeier** [reply]
I'd even replace the term "context" with the term "usage scenario". Time reporting software is obviously used in a different way than games. But even when we measure, say, two slide presentation softwares, the results will heavily depend on the assumed usage. Some people are regularly mailing 20 MB PPT files to 10 persons on cc: around the globe. Shall we factor in such behaviour? It will certainly up the SCI quite a bit. But that is not actually Powerpoints fault. 

In more general terms:
Should we measure the SCI based on average user behaviour or on a recommend usage?

**Reply by atg-abhishek** [reply]
In the example of mailing presentations around, I don't think that should come under the bucket of PowerPoint. What I would imagine instead is that we can propose for some sort of "eco-format" (akin to "eco-mode" that @SaraEmilyBergman talks about) for large-scale mailing or to account for network transmission SCI?

**Reply by atg-abhishek** [reply]
And to the point of accounting for user behaviour, that is another good point @ElmarBorgmeier - I would say that it deserves a new discussion topic of its own. Thoughts?
